Add more Basis implementations for use with EventEffect

TeemuSailynoja opened the pull request:
Add extra Basis functions for the EventEffect.

## Description
Implemented in this PR:

### HalfGaussianBasis
- Add a Gaussian on one side of an event.
- Parameter for creating the effect before/after the event.
- Parameter to control whether the event days should be included or excluded from the effect.

<img width="658" height="601" alt="image" src="https://github.com/user-attachments/assets/fc98f270-621c-4354-ba9a-5a243bb43d56" />

### AsymmetricGaussianBasis
- Adds a Gaussian for both sides of an event
- Parameter for controlling widths and amplitudes before/after the event

<img width="518" height="601" alt="asym-example" src="https://github.com/user-attachments/assets/9766aaab-79d8-404b-9949-1294ed0c8816" />
<img width="518" height="601" alt="asym-example2" src="https://github.com/user-attachments/assets/d918a42a-60e2-4ffc-980b-811acc439042" />

## Related Issue
I jumped straight to implementation.

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [ ] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--1911.org.readthedocs.build/en/1911/

<!-- readthedocs-preview pymc-marketing end -->

cetagostini commented:
Great, @timbo112711 nice if you can add tests and example to the notebook of components!

timbo112711 commented:
@cetagostini looks like @TeemuSailynoja beat me to the tests, much appreciated! I'll add an example to the notebook of components then we should be good here.

timbo112711 commented on pymc_marketing/mmm/events.py:
```diff
@@ -268,6 +268,146 @@ def function(self, x: pt.TensorLike, sigma: pt.TensorLike) -> TensorVariable:
+        a_after: pt.TensorLike,
+    ) -> pt.TensorVariable:
+        """Asymmetric Gaussian bump function."""
+        match self.event_in:
```
@TeemuSailynoja nice modification of the cases here 🙌🏻!

juanitorduz commented:
Nice work @TeemuSailynoja ! Is this one ready for review ?

TeemuSailynoja commented:
Yes. I would like to add a usage example  that would be rendered on the docs page, but the implementation and tests should be ready.

TeemuSailynoja commented:
I just realized two things:

1. As these basis functions are intended to be wrapped in an EventEffect (correct me if this is not the case), which includes an `effect_size` parameter, we should probably have AsymmetricGaussianBasis have one of the amplitudes fixed to one.

2. the `drop` boolean is redundant, as the amplitudes can just be given priors with opposite signs. I could remove that from the class.

timbo112711 commented:
@TeemuSailynoja 

> 1. As these basis functions are intended to be wrapped in an EventEffect (correct me if this is not the case), which includes an `effect_size` parameter, we should probably have AsymmetricGaussianBasis have one of the amplitudes fixed to one. This would avoid a three way unidentifiability between the parameters.

That is correct, this is intended to be wrapped in an `EventEffect`. This is a good point, if we leave it how it is we pose the rise of multiple parameter settings give the same likelihood. Setting one of the amplitudes to 1 will help by

- Having basis just encode the shape asymmetry before vs. after.

- The `effect_size` parameter in `EventEffect` governs the overall scale of the effect.

I agree with you in that having one of the amplitudes fixed to 1, not sure it matters which one however. Maybe we just set the `a_before` to 1. 

> 2. the `drop` boolean is redundant, as the amplitudes can just be given priors with opposite signs.

This makes sense! I'm cool with dropping this arg.

juanitorduz commented:
@TeemuSailynoja, anything to change in this one in view of @PabloRoque 's enhancement https://github.com/pymc-labs/pymc-marketing/pull/1912 ?

TeemuSailynoja commented:
> @TeemuSailynoja, anything to change in this one in view of @PabloRoque 's enhancement #1912 ?

@juanitorduz I changed these implementations to align with #1912 . I have come to agree that that is a better way to implement these :)

I don't see the example plots I added to the new Basis classes in the docs rendered by the readthedocs check/action. Should it be visible there?

juanitorduz commented:
Amazing @TeemuSailynoja !

The plots are not render indeed. Can you try generating the docs locally? Or seeing if there is any little syntax detail that prevents the run? (I can also check tomorrow) :)

TeemuSailynoja commented on pymc_marketing/mmm/events.py:
```diff
@@ -278,10 +278,10 @@ class HalfGaussianBasis(Basis):
         import matplotlib.pyplot as plt
         from pymc_marketing.mmm.events import HalfGaussianBasis
-        from pymc_extras.priors import Prior
+        from pymc_extras.prior import Prior
```
@juanitorduz, I found the issue. A bit of a facepalm moment. Excellent point to call it a night. Fingers crossed we see the plots in the docs now.
